gtsam/inference/Key.h

```diff
@@ -26,13 +26,14 @@
 #include <gtsam/dllexport.h>
 
 #include <boost/function.hpp>
+#include <functional>
```

Comment:
So do we still need boost function header?

Reply:
done.  Side effect: Symbol and LabeledSymbol now need to include `boost/function.hpp` but they arguably should have included this to begin with for exactly this reason because they both explicitly use boost::function.

We should probably go through and replace all boost::function with std::function but probably best for a different PR since this is focused on KeyFormatter

Reply:
We did because some files that included Key.h had a dependency on boost::function.

+1 for @gchenfc's comment.

Reply:
> We did because some files that included Key.h had a dependency on boost::function.
> 
> +1 for @gchenfc's comment.

That's kitchen-sink header thinking. Headers should be included where used, not anywhere else.